src/samples/http/server-sse.c

```diff
@@ -0,0 +1,210 @@
+/*
+ * This file is part of the Soletta Project
+ *
+ * Copyright (C) 2015 Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file
+ * @brief HTTP Server sse
+ *
+ * Sample server that return a response and keep it alive.
+ * It implements server sent events. The content of this
+ * is read from stdin. To see the usage help, -h or --help.
+ */
+
+
+#include <getopt.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include "soletta.h"
+#include "sol-http.h"
+#include "sol-http-server.h"
+#include "sol-util.h"
+#include "sol-util-file.h"
+
+static struct sol_http_server *server;
+static struct sol_fd *stdin_watch;
+static struct sol_buffer value = SOL_BUFFER_INIT_EMPTY;
+static struct sol_ptr_vector responses = SOL_PTR_VECTOR_INIT;
+
+static bool
+on_stdin(void *data, int fd, uint32_t flags)
+{
+    uint16_t i;
+    struct sol_http_response_sse *sse;
+
+    if (flags & (SOL_FD_FLAGS_ERR | SOL_FD_FLAGS_HUP)) {
+        fprintf(stderr, "ERROR: Something wrong happened with file descriptor: %d\n", fd);
+        goto err;
+    }
+
+    if (flags & SOL_FD_FLAGS_IN) {
+        int err;
+
+        value.used = 0;
+        err = sol_buffer_insert_bytes(&value, 0,
+            (uint8_t *)"data: ", sizeof("data: ") - 1);
+        if (err < 0) {
+            fprintf(stderr, "ERROR: failed to append the data prefix: %s\n",
+                sol_util_strerrora(-err));
+            goto err;
+        }
+        /* this will loop trying to read as much data as possible to buffer. */
+        err = sol_util_load_file_fd_buffer(fd, &value);
```

Comment:
this is a sample, i guess this will bring more complexity than necessary to understand it. The normal usage of this will be run from terminal in canonical mode, it meas that the data will come when return is pressed (no \n).  To check the \n properly, it's necessary split the string into tokens then prepend `data: ` and append `\n\n`.